File: IFVG_Realtime.py
"""IFVG Signals Realtime (No Wait) - Python translation.

This module mirrors the Pine Script logic in `S1_Silver_Bullet/IFVG_Realtime.txt` by:
- Scanning for bullish/bearish FVGs within a configurable lookback window.
- Validating minimum FVG size and checking for prior breaks in confirmed bars.
- Emitting realtime buy/sell signals when price crosses the FVG boundary.
- Applying an EMA/SMA slope + price filter at the time of signal.

The implementation iterates bar-by-bar to reflect the Pine behavior where
conditions are evaluated using the current close without waiting for bar close.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Tuple

import pandas as pd


@dataclass
class IFVGZone:
    direction: str  # "buy" or "sell"
    left_index: int
    right_index: int
    top: float
    bottom: float


def _sma(series: pd.Series, length: int) -> pd.Series:
    return series.rolling(length, min_periods=length).mean()


def _ema(series: pd.Series, length: int) -> pd.Series:
    return series.ewm(span=length, adjust=False, min_periods=length).mean()


def _detect_fvg(
    df: pd.DataFrame,
    current_index: int,
    offset: int,
    eps_val: float,
) -> int:
    """Return 1 for bull gap, -1 for bear gap, 0 for none."""
    lookback_index = current_index - (offset + 2)
    if lookback_index < 0:
        return 0

    h2 = float(df["high"].iloc[lookback_index])
    l2 = float(df["low"].iloc[lookback_index])
    lt = float(df["low"].iloc[current_index - offset])
    ht = float(df["high"].iloc[current_index - offset])

    if lt > h2 - eps_val:
        return 1
    if ht < l2 + eps_val:
        return -1
    return 0


def compute_ifvg_realtime(
    df: pd.DataFrame,
    *,
    mintick: float,
    pip_size_multiplier: float = 1.0,
    ifvg_gap_bars: int = 15,
    min_fvg_pips: float = 0.0,
    fvg_eps_points: float = 0.0,
    show_zones: bool = True,
    ma_period: int = 21,
    ma_kind: str = "EMA",
) -> Tuple[pd.DataFrame, List[IFVGZone]]:
    """Compute IFVG realtime signals.

    Args:
        df: DataFrame with columns: open, high, low, close.
        mintick: Minimum tick size (syminfo.mintick in Pine).
        pip_size_multiplier: Multiplier for tick size (kept for parity).
        ifvg_gap_bars: Lookback bars for FVG search.
        min_fvg_pips: Minimum FVG size in pips/points.
        fvg_eps_points: Epsilon tolerance in price units.
        show_zones: Whether to collect IFVG zones.
        ma_period: Moving average period.
        ma_kind: "EMA" or "SMA".

    Returns:
        Tuple of (DataFrame with signal columns, list of IFVGZone entries).
    """
    if ma_kind not in {"EMA", "SMA"}:
        raise ValueError("ma_kind must be 'EMA' or 'SMA'.")

    price = df["close"].astype(float)
    ma_val = _ema(price, ma_period) if ma_kind == "EMA" else _sma(price, ma_period)

    _ = mintick * pip_size_multiplier  # retained for parity with Pine settings

    min_size_val = min_fvg_pips * (mintick * 10)

    signal_dir: List[int] = []
    zones: List[IFVGZone] = []

    for idx in range(len(df)):
        current_signal = 0

        for i in range(1, ifvg_gap_bars + 1):
            if idx - (i + 2) < 0:
                continue

            fvg_type = _detect_fvg(df, idx, i, fvg_eps_points)

            if fvg_type == 1:
                gap_low = float(df["high"].iloc[idx - (i + 2)])
                gap_high = float(df["low"].iloc[idx - i])

                if (gap_high - gap_low) >= min_size_val:
                    already_broken = False
                    if i > 1:
                        for k in range(i - 1, 0, -1):
                            if float(df["close"].iloc[idx - k]) < gap_low:
                                already_broken = True
                                break

                    if not already_broken and float(df["close"].iloc[idx]) < gap_low:
                        ma_condition = False
                        if idx - 1 >= 0 and pd.notna(ma_val.iloc[idx]) and pd.notna(ma_val.iloc[idx - 1]):
                            ma_condition = ma_val.iloc[idx] < ma_val.iloc[idx - 1] and float(
                                df["close"].iloc[idx]
                            ) < float(ma_val.iloc[idx])

                        if ma_condition:
                            current_signal = -1
                            if show_zones:
                                zones.append(
                                    IFVGZone(
                                        direction="sell",
                                        left_index=idx - (i + 2),
                                        right_index=idx,
                                        top=gap_high,
                                        bottom=gap_low,
                                    )
                                )
                            break

            if fvg_type == -1:
                gap_low2 = float(df["high"].iloc[idx - i])
                gap_high2 = float(df["low"].iloc[idx - (i + 2)])

                if (gap_high2 - gap_low2) >= min_size_val:
                    already_broken2 = False
                    if i > 1:
                        for k in range(i - 1, 0, -1):
                            if float(df["close"].iloc[idx - k]) > gap_high2:
                                already_broken2 = True
                                break

                    if not already_broken2 and float(df["close"].iloc[idx]) > gap_high2:
                        ma_condition2 = False
                        if idx - 1 >= 0 and pd.notna(ma_val.iloc[idx]) and pd.notna(ma_val.iloc[idx - 1]):
                            ma_condition2 = ma_val.iloc[idx] > ma_val.iloc[idx - 1] and float(
                                df["close"].iloc[idx]
                            ) > float(ma_val.iloc[idx])

                        if ma_condition2:
                            current_signal = 1
                            if show_zones:
                                zones.append(
                                    IFVGZone(
                                        direction="buy",
                                        left_index=idx - (i + 2),
                                        right_index=idx,
                                        top=gap_high2,
                                        bottom=gap_low2,
                                    )
                                )
                            break

        signal_dir.append(current_signal)

    result = df.copy()
    result["signal_dir"] = signal_dir
    result["buy_signal"] = result["signal_dir"] == 1
    result["sell_signal"] = result["signal_dir"] == -1
    result["ma"] = ma_val

    return result, zones

File: IFVG_Realtime.cs
using System;
using cAlgo.API;
using cAlgo.API.Indicators;

namespace cAlgo
{
    [Indicator(IsOverlay = true, TimeZone = TimeZones.UTC, AccessRights = AccessRights.None)]
    public class IFVGRealtime : Indicator
    {
        [Parameter("FVG Search Lookback (Bars)", DefaultValue = 15, MinValue = 1)]
        public int FvgGapBars { get; set; }

        [Parameter("Min FVG Size (Pips/Points)", DefaultValue = 0.0)]
        public double MinFvgPips { get; set; }

        [Parameter("FVG Epsilon (Price Units)", DefaultValue = 0.0)]
        public double FvgEpsilonPoints { get; set; }

        [Parameter("Show IFVG Zones", DefaultValue = true)]
        public bool ShowZones { get; set; }

        [Parameter("Zone Color Buy", DefaultValue = "#062625")]
        public Color ZoneColorBuy { get; set; }

        [Parameter("Zone Color Sell", DefaultValue = "#3E003E")]
        public Color ZoneColorSell { get; set; }

        [Parameter("MA Period", DefaultValue = 21, MinValue = 1)]
        public int MaPeriod { get; set; }

        [Parameter("MA Type", DefaultValue = "EMA")]
        public string MaType { get; set; }

        private const string BuySignalColorHex = "#008080"; // teal
        private const string SellSignalColorHex = "#800000"; // maroon
        private const string BuyAlertMessage = "IFVG Buy Signal (Realtime) | IFVG Signal Triggered";
        private const string SellAlertMessage = "IFVG Sell Signal (Realtime) | IFVG Signal Triggered";

        private IndicatorDataSeries _maSeries;
        private SimpleMovingAverage _sma;
        private ExponentialMovingAverage _ema;
        private int _lastAlertBar = -1;
        private int _lastAlertDir;

        protected override void Initialize()
        {
            _maSeries = CreateDataSeries();
            _sma = Indicators.SimpleMovingAverage(Bars.ClosePrices, MaPeriod);
            _ema = Indicators.ExponentialMovingAverage(Bars.ClosePrices, MaPeriod);
        }

        public override void Calculate(int index)
        {
            var minSizeValue = MinFvgPips * Symbol.PipSize;
            var calcEps = FvgEpsilonPoints;

            var maVal = CalculateMa(index);
            var signalDir = 0;

            RemoveRealtimeZoneObjects(index);

            for (var i = 1; i <= FvgGapBars; i++)
            {
                var fvgType = DetectFvg(index, i, calcEps);
                if (fvgType == 0)
                    continue;

                if (TryProcessFvgCandidate(index, i, fvgType, minSizeValue, maVal, out signalDir))
                    break;
            }

            DrawSignals(index, signalDir);
            EmitAlerts(index, signalDir);
        }

        private bool TryProcessFvgCandidate(int index, int i, int fvgType, double minSizeValue, double maVal, out int signalDir)
        {
            signalDir = 0;

            var isBearishGap = fvgType == 1;
            var gapLow = isBearishGap ? Bars.HighPrices[index - (i + 2)] : Bars.HighPrices[index - i];
            var gapHigh = isBearishGap ? Bars.LowPrices[index - i] : Bars.LowPrices[index - (i + 2)];

            if ((gapHigh - gapLow) < minSizeValue)
                return false;

            var alreadyBroken = false;
            if (i > 1)
            {
                for (var k = i - 1; k >= 1; k--)
                {
                    var close = Bars.ClosePrices[index - k];
                    if ((isBearishGap && close < gapLow) || (!isBearishGap && close > gapHigh))
                    {
                        alreadyBroken = true;
                        break;
                    }
                }
            }

            if (alreadyBroken)
                return false;

            var breakout = isBearishGap ? Bars.ClosePrices[index] < gapLow : Bars.ClosePrices[index] > gapHigh;
            if (!breakout)
                return false;

            var maReady = !double.IsNaN(maVal) && !double.IsNaN(_maSeries[index - 1]);
            var maCondition = isBearishGap
                ? maReady && maVal < _maSeries[index - 1] && Bars.ClosePrices[index] < maVal
                : maReady && maVal > _maSeries[index - 1] && Bars.ClosePrices[index] > maVal;

            if (!maCondition)
                return false;

            signalDir = isBearishGap ? -1 : 1;
            if (ShowZones)
                DrawIfvgZone(index, i, gapHigh, gapLow, !isBearishGap);

            return true;
        }

        private int DetectFvg(int currentIndex, int idx, double epsVal)
        {
            if (idx + 2 > currentIndex)
                return 0;

            var h2 = Bars.HighPrices[currentIndex - (idx + 2)];
            var l2 = Bars.LowPrices[currentIndex - (idx + 2)];
            var lt = Bars.LowPrices[currentIndex - idx];
            var ht = Bars.HighPrices[currentIndex - idx];

            if (lt > h2 - epsVal)
                return 1;

            if (ht < l2 + epsVal)
                return -1;

            return 0;
        }

        private double CalculateMa(int index)
        {
            if (string.Equals(MaType, "SMA", StringComparison.OrdinalIgnoreCase))
                _maSeries[index] = _sma.Result[index];
            else
                _maSeries[index] = _ema.Result[index];

            return _maSeries[index];
        }

        private void DrawIfvgZone(int index, int i, double top, double bottom, bool isBuy)
        {
            var leftIndex = index - (i + 2);
            var zoneId = $"ifvg_zone_{index}_{i}_{(isBuy ? "buy" : "sell")}";
            var baseColor = isBuy ? ZoneColorBuy : ZoneColorSell;
            var zoneColor = Color.FromArgb(153, baseColor.R, baseColor.G, baseColor.B);

            var rect = Chart.DrawRectangle(zoneId, leftIndex, top, index, bottom, zoneColor);
            rect.IsFilled = true;
            rect.IsInteractive = false;
        }

        private void RemoveRealtimeZoneObjects(int index)
        {
            for (var i = 1; i <= FvgGapBars; i++)
            {
                Chart.RemoveObject($"ifvg_zone_{index}_{i}_buy");
                Chart.RemoveObject($"ifvg_zone_{index}_{i}_sell");
            }
        }

        private void DrawSignals(int index, int signalDir)
        {
            var buyId = $"ifvg_buy_{index}";
            var sellId = $"ifvg_sell_{index}";

            if (signalDir == 1)
                Chart.DrawIcon(buyId, ChartIconType.UpTriangle, index, Bars.LowPrices[index], Color.FromHex(BuySignalColorHex));

            if (signalDir == -1)
                Chart.DrawIcon(sellId, ChartIconType.DownTriangle, index, Bars.HighPrices[index], Color.FromHex(SellSignalColorHex));
        }

        private void EmitAlerts(int index, int signalDir)
        {
            if (index != Bars.Count - 1)
                return;

            if (signalDir != 0 && (_lastAlertBar != index || _lastAlertDir != signalDir))
            {
                Print(signalDir == 1 ? BuyAlertMessage : SellAlertMessage);
                _lastAlertBar = index;
                _lastAlertDir = signalDir;
            }
        }
    }
}
